### PR TITLE
Go modules

### PR DESCRIPTION
**Pull request by tzdybal**
Go 1.13, the next release of Go, will enable modules by default.
This PR introduces Go modules into go-ethereum. 

`env.sh` script was removed, as it's not needed anymore. The behavior of build process remains the same - project can be build from any directory, inside or outside of `$GOPATH`.

### CODE REVIEW

**Comment by hadv** [comment]
We can remove the vendor folder also if we switch to use go mod, right?

**Comment by i-norden** [comment]
I'm curious if you have tried this update closer to the head state? go mod inititalization for me at 7f3362595a3bdf1c27b17f16b7d90d89882312ab throws unknown revision errors for `usb` and `protobuf` dependencies, which have both changed since this was opened. 

errors:
`go: github.com/karalabe/usb@v0.0.0-20190703132631-6a7de9d893fe: unknown revision 6a7de9d893fe` 

and

`github.com/gogo/protobuf@v0.0.0-00010101000000-000000000000: unknown revision 000000000000`

In the case of karalable/usb the issue appears to be exactly what it says, there is no revision 6a7de9d893fe in the [repo](https://github.com/karalabe/usb/search?q=6a7de9d893fe&type=Commits) anymore. There was on another branch but that was [merged in](https://github.com/karalabe/usb/pull/4) as [9be757f](https://github.com/karalabe/usb/commit/9be757f914c0907b7ddd561ea86eec15313ac022). I will open an issue/PR for this unless I realize I'm missing something... 

Not sure what the deal is with protobuf though, as the revision we should be using at 7f3362595a3bdf1c27b17f16b7d90d89882312ab is [b285ee9cfc6c881bb20c0d8dc73370ea9b9ec90f ](https://github.com/ethereum/go-ethereum/blob/master/vendor/vendor.json#L140) which [does exist](https://github.com/golang/protobuf/search?q=b285ee9cfc6c881bb20c0d8dc73370ea9b9ec90f&type=Commits) at that repo, but for some reason it is trying to use the github.com/gogo/protobuf repo which does not have that revision (even though my go.mod has `github.com/golang/protobuf v1.3.1`).

**Comment by tzdybal** [comment]
I'll re-approach this again in few days. I want to fix problem with linting, and ofc all the errors related to `go mod`.

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	github.com/cespare/cp v0.1.0
+	github.com/codahale/hdrhistogram v0.0.0-20161010025455-3a0bb77429bd // indirect
+	github.com/davecgh/go-spew v1.1.1
+	github.com/deckarep/golang-set v0.0.0-20180603214616-504e848d77ea
+	github.com/docker/docker v0.0.0-20180625184442-8e610b2b55bf
+	github.com/edsrzf/mmap-go v0.0.0-20160512033002-935e0e8a636c
+	github.com/elastic/gosigar v0.0.0-20180330100440-37f05ff46ffa
+	github.com/fatih/color v1.6.0
+	github.com/fjl/memsize v0.0.0-20180418122429-ca190fb6ffbc
+	github.com/gballet/go-libpcsclite v0.0.0-20190403181518-312b5175032f
```
Comment (gballet) [inline go.mod]: This should be upgraded to the latest version, which is `2772fd86a8ff4306d2749f610a386bfee9e0d727`

File: accounts/scwallet/hub.go [hunk 2]
```diff
@@ -153,7 +153,7 @@ func (hub *Hub) setPairing(wallet *Wallet, pairing *smartcardPairing) error {
 
 // NewHub creates a new hardware wallet manager for smartcards.
 func NewHub(daemonPath string, scheme string, datadir string) (*Hub, error) {
-	context, err := pcsc.EstablishContext(daemonPath, pcsc.ScopeSystem)
+	context, err := pcsc.EstablishContext(pcsc.ScopeSystem)
```
Comment (gballet) [inline accounts/scwallet/hub.go]: You went back to an older version of pcsc, this will break features.

Comment (tzdybal) [reply accounts/scwallet/hub.go]: I did it by accident, now it's fixed.

**Comment by tzdybal** [comment]
All linting errors resolved. Rebased with master. Build passing.

**Comment by karalabe** [comment]
Copied this from Discord:

The theory is easy. Unfortunately there are certain builders that will choke on modules
The most notorious is Launchpad PPAs, because they don't have Go 1.13 available as official packages
And there's also no 3rd party PPA with updated Go for all Ubuntu releases (there was one, but was discontinued due to some build issues on PPA, ironic)
I've tried 2 solutions already, one on Friday (download Go 1.13 during PPA build) and one on Monday (upload binaries directly) both failed (chroot with net disabled, PPA must build it itself)
As such now we need to set up an infra to build the packages via Snaps first, and then hack in an empty-package dependency on PPA
Until that's done, we can't enable modules as they would break the binary distributions

**Comment by tzdybal** [comment]
If modules cannot be used at all, I assume only go 1.10 (or prior) is available (which is rather old and therefore should be updated anyways).

It seems that [longsleep/golang-backports PPA](https://launchpad.net/~longsleep/+archive/ubuntu/golang-backports/+packages) provides up-to-date packages for all currently supported Ubuntu distributions (see [Ubuntu release page](https://wiki.ubuntu.com/Releases)).

Is it really necessary to build packages for unsupported versions of Ubuntu?

**Comment by karalabe** [comment]
Closing this in favor of https://github.com/ethereum/go-ethereum/pull/20311

- Still works with Go 1.11 and 1.12, which we cannot drop due to Ubuntu Trusty support.
- Fixes the issue of testing the `bind` package with the correct modules (without generating temp files within the live source tree).
- Implements Go module support for PPA builds in Launchpad.
- Ensures build work through the `xgo` cross compiler for MIPS.
- Generally just focuses on go module support and not on linter fixes.

**Pull request closed without merging**